server/src/test/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhaseTests.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.fetch.subphase;
+
+import org.elasticsearch.action.search.SearchPhaseExecutionException;
+import org.elasticsearch.action.search.SearchRequestBuilder;
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.action.support.WriteRequest;
+import org.elasticsearch.common.document.DocumentField;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentFactory;
+import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.test.ESSingleNodeTestCase;
+import org.junit.Before;
+
+import java.io.IOException;
+import java.util.Map;
+
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.hasItems;
+
+public class FetchFieldsPhaseTests extends ESSingleNodeTestCase {
```

Comment:
I tried hard to write proper unit tests (in the style of `FetchSourcePhaseTests`), but it's currently challenging to stub out the required pieces like source lookup.

Reply:
What if you made something like:
```
public void hitsExecute(Function<SearchHit, Map<String, Object>> sourceLookup, SearchHit[] hits)
```

and unit tested that? Then and the only thing missing a unit test would be the map from search context to "simple" classes?

Reply:
I had been hesitant to do this, because I thought it would make `FetchFieldsPhase` harder to read... but I actually think it turned out pretty cleanly!